### PR TITLE
Copr: vectorize sin, cos, tan, cot

### PR DESCRIPTION
**Pull request by koushiro**
PCP #5751

Signed-off-by: koushiro <koushiro.cqx@gmail.com>

###  What have you changed?

Vectorize `sin`, `cos`, `tan` and `cot`

###  What is the type of the changes?

- Improvement (a change which is an improvement to an existing feature)

###  How is the PR tested?

- Unit test
- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 1]
```diff
@@ -253,21 +253,57 @@ fn sqrt(arg: &Option<Real>) -> Result<Option<Real>> {
     }))
 }
 
+#[inline]
+#[rpn_fn]
+fn sin(arg: &Option<Real>) -> Result<Option<Real>> {
+    Ok(arg.map(|arg| Real::from(arg.sin())))
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: I would prefer to use `Real::new()` so that invalid values are mapped to `None` instead of panic.

Comment (koushiro) [reply components/tidb_query/src/rpn_expr/impl_math.rs]: Okay, that sounds better

File: components/tidb_query/src/rpn_expr/impl_math.rs [hunk 2]
```diff
@@ -253,21 +253,57 @@ fn sqrt(arg: &Option<Real>) -> Result<Option<Real>> {
...
+
+#[inline]
+#[rpn_fn(capture = [ctx])]
+fn cot(ctx: &mut EvalContext, arg: &Option<Real>) -> Result<Option<Real>> {
+    match arg {
+        Some(arg) => {
+            let tan = arg.tan();
+            if tan != 0.0 {
+                let cot = 1.0 / tan;
+                if !cot.is_infinite() && !cot.is_nan() {
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_math.rs]: `!cot.is_nan()` can be eliminated since `Real::new()` will check it :)

**Comment by lonng** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/29

**Comment by breeswish** [comment]
Looks like `cot(0)` error behaviour is not the same as TiDB: https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tikv_ghpr_integration-copr-test/detail/tikv_ghpr_integration-copr-test/877/

**Comment by koushiro** [comment]
> Looks like `cot(0)` error behaviour is not the same as TiDB: https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/tikv_ghpr_integration-copr-test/detail/tikv_ghpr_integration-copr-test/877/

Running tests locally seems to be fine

<details>
  <summary>include=5_math_1.sql make push-down-test</summary>

```
cd push-down-test && bash run-tests.sh

+ Cleaning up temp directory stale data: /tmp/copr_test

+ Killing existing tidb / tikv / pd process
tidb-server: no process found
tikv-server: no process found
pd-server: no process found

+ Building TiDB
  - Building TiDB binary with failpoint enabled from /home/koushiro/Code/gocode/src/github.com/pingcap/tidb
make[1]: Entering directory '/home/koushiro/Code/gocode/src/github.com/pingcap/tidb'
make[1]: Leaving directory '/home/koushiro/Code/gocode/src/github.com/pingcap/tidb'
make[1]: Entering directory '/home/koushiro/Code/gocode/src/github.com/pingcap/tidb'
CGO_ENABLED=1 GO111MODULE=on go build  -tags codes -trimpath  -ldflags '-X "github.com/pingcap/parser/mysql.TiDBReleaseVersion=v4.0.0-alpha-856-g65389a164-dirty" -X "github.com/pingcap/tidb/util/printer.TiDBBuildTS=2019-11-14 03:51:53" -X "github.com/pingcap/tidb/util/printer.TiDBGitHash=65389a1644c3e1cff182186a786bf5b0c24a514a" -X "github.com/pingcap/tidb/util/printer.TiDBGitBranch=master" -X "github.com/pingcap/tidb/util/printer.GoVersion=go version go1.13.4 linux/amd64" ' -o bin/tidb-server tidb-server/main.go
Build TiDB Server successfully!
make[1]: Leaving directory '/home/koushiro/Code/gocode/src/github.com/pingcap/tidb'

+ Building Push Down Tester
  - Building from ./bin/push_down_test_bin

+ PD version
Release Version: v4.0.0-alpha-158-g099369e5
Git Commit Hash: 099369e539bf12a96662fb9d5bf6776f3c9b3c58
Git Branch: master
UTC Build Time:  2019-11-14 03:06:41

+ Starting PD
  - Sleep 3s to wait for PD to start

+ TiKV version
TiKV 
Release Version:   4.0.0-alpha
Git Commit Hash:   Unknown (env var does not exist when building)
Git Commit Branch: Unknown (env var does not exist when building)
UTC Build Time:    Unknown (env var does not exist when building)
Rust Version:      Unknown (env var does not exist when building)

+ Starting TiKV
  - Sleep 3s to wait for TiKV to start

+ TiDB version
2019-11-14 11:52:05.644891 I | maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined
Release Version: v4.0.0-alpha-856-g65389a164-dirty
Git Commit Hash: 65389a1644c3e1cff182186a786bf5b0c24a514a
Git Branch: master
UTC Build Time: 2019-11-14 03:51:53
GoVersion: go version go1.13.4 linux/amd64
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

+ Starting TiDB
  - Sleep 10s to wait for TiDB to start
2019-11-14 11:52:05.657482 I | maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined
2019-11-14 11:52:05.657531 I | maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined
2019-11-14 11:52:05.658627 I | maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined

+ Waiting TiDB start up
  - Waiting TiDB (no push down)
+--------------------+
| Database           |
+--------------------+
| INFORMATION_SCHEMA |
| PERFORMANCE_SCHEMA |
| mysql              |
| test               |
+--------------------+
  - TiDB startup successfully (no push down)
  - Waiting TiDB (push down without vectorization)
+--------------------+
| Database           |
+--------------------+
| INFORMATION_SCHEMA |
| PERFORMANCE_SCHEMA |
| mysql              |
| test               |
+--------------------+
  - TiDB startup successfully (push down without vectorization)
  - Waiting TiDB (push down with vectorization)
+--------------------+
| Database           |
+--------------------+
| INFORMATION_SCHEMA |
| PERFORMANCE_SCHEMA |
| mysql              |
| test               |
+--------------------+
  - TiDB startup successfully (push down with vectorization)

+ Test Configurations
  - tidb_no_push_down_pid=45995
  - tikv_no_batch_pid=45790
  - pd_no_batch_pid=45746
  - tidb_push_down_no_batch_pid=45996
  - tikv_with_batch_pid=45791
  - pd_with_batch_pid=45747
  - tidb_push_down_with_batch_pid=45997

+ Start test
2019/11/14 11:52:15 Preparing database [push_down_test_db] for [root@tcp(127.0.0.1:4005)/{db}?allowNativePasswords=true]...
2019/11/14 11:52:15 Preparing database [push_down_test_db] for [root@tcp(127.0.0.1:4006)/{db}?allowNativePasswords=true]...
2019/11/14 11:52:15 Preparing database [push_down_test_db] for [root@tcp(127.0.0.1:4007)/{db}?allowNativePasswords=true]...
2019/11/14 11:52:16 Serial Testing [prepare/0_data.sql]...
2019/11/14 11:52:16 Parsing...[prepare/0_data.sql]
2019/11/14 11:52:16 Running...[prepare/0_data.sql]
2019/11/14 11:52:21 2019/11/14 11:52:21 Test summary: non-matching queries: 0, success queries: 67, skipped queries: 0
2019/11/14 11:52:21 Test summary(prepare/0_data.sql): Test case PASS

2019/11/14 11:52:21 All test finished: pass cases: 1, fail cases: 0
2019/11/14 11:52:21 Prepare finished, start testing...
2019/11/14 11:52:21 Parallel Testing [sql/randgen/5_math_1.sql]...
2019/11/14 11:52:21 Parsing...[sql/randgen/5_math_1.sql]
2019/11/14 11:52:21 Running...[sql/randgen/5_math_1.sql]
2019/11/14 11:52:38 2019/11/14 11:52:38 Test summary: non-matching queries: 0, success queries: 886, skipped queries: 114
2019/11/14 11:52:38 Test summary(sql/randgen/5_math_1.sql): Test case PASS

2019/11/14 11:52:38 All test finished: pass cases: 1, fail cases: 0
+ Test finished
  - ./bin/push_down_test_bin exit code is 0

+ Killing existing tidb / tikv / pd process
run-tests.sh: line 138: 45996 Killed                  "$tidb_bin" -log-file "$push_down_no_batch_tidb_log_file" -config "$push_down_no_batch_config_dir"/tidb.toml -L ${log_level}
```
</details>

<details>
  <summary>include=5_math_2.sql make push-down-test</summary>

```
cd push-down-test && bash run-tests.sh

+ Cleaning up temp directory stale data: /tmp/copr_test

+ Killing existing tidb / tikv / pd process
tidb-server: no process found
tikv-server: no process found
pd-server: no process found

+ Building TiDB
  - Building TiDB binary with failpoint enabled from /home/koushiro/Code/gocode/src/github.com/pingcap/tidb
make[1]: Entering directory '/home/koushiro/Code/gocode/src/github.com/pingcap/tidb'
make[1]: Leaving directory '/home/koushiro/Code/gocode/src/github.com/pingcap/tidb'
make[1]: Entering directory '/home/koushiro/Code/gocode/src/github.com/pingcap/tidb'
CGO_ENABLED=1 GO111MODULE=on go build  -tags codes -trimpath  -ldflags '-X "github.com/pingcap/parser/mysql.TiDBReleaseVersion=v4.0.0-alpha-856-g65389a164-dirty" -X "github.com/pingcap/tidb/util/printer.TiDBBuildTS=2019-11-14 03:49:59" -X "github.com/pingcap/tidb/util/printer.TiDBGitHash=65389a1644c3e1cff182186a786bf5b0c24a514a" -X "github.com/pingcap/tidb/util/printer.TiDBGitBranch=master" -X "github.com/pingcap/tidb/util/printer.GoVersion=go version go1.13.4 linux/amd64" ' -o bin/tidb-server tidb-server/main.go
Build TiDB Server successfully!
make[1]: Leaving directory '/home/koushiro/Code/gocode/src/github.com/pingcap/tidb'

+ Building Push Down Tester
  - Building from ./bin/push_down_test_bin

+ PD version
Release Version: v4.0.0-alpha-158-g099369e5
Git Commit Hash: 099369e539bf12a96662fb9d5bf6776f3c9b3c58
Git Branch: master
UTC Build Time:  2019-11-14 03:06:41

+ Starting PD
  - Sleep 3s to wait for PD to start

+ TiKV version
TiKV 
Release Version:   4.0.0-alpha
Git Commit Hash:   Unknown (env var does not exist when building)
Git Commit Branch: Unknown (env var does not exist when building)
UTC Build Time:    Unknown (env var does not exist when building)
Rust Version:      Unknown (env var does not exist when building)

+ Starting TiKV
  - Sleep 3s to wait for TiKV to start

+ TiDB version
2019-11-14 11:50:11.464182 I | maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined
Release Version: v4.0.0-alpha-856-g65389a164-dirty
Git Commit Hash: 65389a1644c3e1cff182186a786bf5b0c24a514a
Git Branch: master
UTC Build Time: 2019-11-14 03:49:59
GoVersion: go version go1.13.4 linux/amd64
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false

+ Starting TiDB
  - Sleep 10s to wait for TiDB to start
2019-11-14 11:50:11.476138 I | maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined
2019-11-14 11:50:11.476357 I | maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined
2019-11-14 11:50:11.477305 I | maxprocs: Leaving GOMAXPROCS=8: CPU quota undefined

+ Waiting TiDB start up
  - Waiting TiDB (no push down)
+--------------------+
| Database           |
+--------------------+
| INFORMATION_SCHEMA |
| PERFORMANCE_SCHEMA |
| mysql              |
| test               |
+--------------------+
  - TiDB startup successfully (no push down)
  - Waiting TiDB (push down without vectorization)
+--------------------+
| Database           |
+--------------------+
| INFORMATION_SCHEMA |
| PERFORMANCE_SCHEMA |
| mysql              |
| test               |
+--------------------+
  - TiDB startup successfully (push down without vectorization)
  - Waiting TiDB (push down with vectorization)
+--------------------+
| Database           |
+--------------------+
| INFORMATION_SCHEMA |
| PERFORMANCE_SCHEMA |
| mysql              |
| test               |
+--------------------+
  - TiDB startup successfully (push down with vectorization)

+ Test Configurations
  - tidb_no_push_down_pid=45295
  - tikv_no_batch_pid=45090
  - pd_no_batch_pid=45049
  - tidb_push_down_no_batch_pid=45296
  - tikv_with_batch_pid=45091
  - pd_with_batch_pid=45050
  - tidb_push_down_with_batch_pid=45297

+ Start test
2019/11/14 11:50:21 Preparing database [push_down_test_db] for [root@tcp(127.0.0.1:4005)/{db}?allowNativePasswords=true]...
2019/11/14 11:50:21 Preparing database [push_down_test_db] for [root@tcp(127.0.0.1:4006)/{db}?allowNativePasswords=true]...
2019/11/14 11:50:21 Preparing database [push_down_test_db] for [root@tcp(127.0.0.1:4007)/{db}?allowNativePasswords=true]...
2019/11/14 11:50:21 Serial Testing [prepare/0_data.sql]...
2019/11/14 11:50:21 Parsing...[prepare/0_data.sql]
2019/11/14 11:50:21 Running...[prepare/0_data.sql]
2019/11/14 11:50:26 2019/11/14 11:50:26 Test summary: non-matching queries: 0, success queries: 67, skipped queries: 0
2019/11/14 11:50:26 Test summary(prepare/0_data.sql): Test case PASS

2019/11/14 11:50:26 All test finished: pass cases: 1, fail cases: 0
2019/11/14 11:50:26 Prepare finished, start testing...
2019/11/14 11:50:26 Parallel Testing [sql/randgen/5_math_2.sql]...
2019/11/14 11:50:26 Parsing...[sql/randgen/5_math_2.sql]
2019/11/14 11:50:26 Running...[sql/randgen/5_math_2.sql]
2019/11/14 11:51:17 2019/11/14 11:51:17 Test summary: non-matching queries: 0, success queries: 2734, skipped queries: 262
2019/11/14 11:51:17 Test summary(sql/randgen/5_math_2.sql): Test case PASS

2019/11/14 11:51:17 All test finished: pass cases: 1, fail cases: 0
+ Test finished
  - ./bin/push_down_test_bin exit code is 0

+ Killing existing tidb / tikv / pd process
run-tests.sh: line 138: 45295 Killed                  "$tidb_bin" -log-file "$no_push_down_tidb_log_file" -config "$no_push_down_config_dir"/tidb.toml -L ${log_level}
run-tests.sh: line 138: 45296 Killed                  "$tidb_bin" -log-file "$push_down_no_batch_tidb_log_file" -config "$push_down_no_batch_config_dir"/tidb.toml -L ${log_level}
run-tests.sh: line 138: 45297 Killed                  "$tidb_bin" -log-file "$push_down_with_batch_tidb_log_file" -config "$push_down_with_batch_config_dir"/tidb.toml -L ${log_level}
```
</details>

**Comment by cireu** [comment]
```diff
diff --git a/components/tidb_query/src/rpn_expr/impl_math.rs b/components/tidb_query/src/rpn_expr/impl_math.rs
index 8abe3066..3d7ddc47 100644
--- a/components/tidb_query/src/rpn_expr/impl_math.rs
+++ b/components/tidb_query/src/rpn_expr/impl_math.rs
@@ -272,8 +272,8 @@ fn tan(arg: &Option<Real>) -> Result<Option<Real>> {
 }
 
 #[inline]
-#[rpn_fn(capture = [ctx])]
-fn cot(ctx: &mut EvalContext, arg: &Option<Real>) -> Result<Option<Real>> {
+#[rpn_fn]
+fn cot(arg: &Option<Real>) -> Result<Option<Real>> {
     match arg {
         Some(arg) => {
             let tan = arg.tan();
@@ -283,9 +283,7 @@ fn cot(ctx: &mut EvalContext, arg: &Option<Real>) -> Result<Option<Real>> {
                     return Ok(Real::new(cot).ok());
                 }
             }
-            Ok(ctx
-                .handle_overflow_err(Error::overflow("DOUBLE", format!("cot({})", arg)))
-                .map(|_| None)?)
+            Err(Error::overflow("DOUBLE", format!("cot({})", arg)).into())
         }
         None => Ok(None),
     }
```

this can keep the behaviour consistency between TiDB and TiKV, but seems to break integration test 33 @breeswish  :thinking:

**Comment by breeswish** [comment]
I guess @cireu 's code is correct since this is the only difference.

**Comment by koushiro** [comment]
@sre-bot /run-integration-copr-test copr-test=pr/29

**Comment by breeswish** [comment]
Please fix conflicts. Thanks!

**Comment by lonng** [comment]
@koushiro Please resolve the conflicts.

**Pull request merged**